File: backend/src/techinterview/core/schemas.py
"""共用 Pydantic schema（T020）。

對外 JSON 一律 camelCase（以 alias 產生），資料庫欄位為 snake_case。
"""

from __future__ import annotations

from enum import StrEnum

from pydantic import BaseModel, ConfigDict, Field, field_validator
from pydantic.alias_generators import to_camel

# 作答內容上限 256 KB（data-model.md：answer 驗證規則）
MAX_CONTENT_BYTES = 256 * 1024

# 平台外工具事件門檻：短於 1000ms 的離開 MUST NOT 記錄
ENVIRONMENT_EVENT_MIN_DURATION_MS = 1000

# 剩餘時間低於此值時計時器轉為警示呈現（FR-020）
TIMER_WARNING_THRESHOLD_SEC = 5 * 60


class Language(StrEnum):
    JAVASCRIPT = "javascript"
    TYPESCRIPT = "typescript"
    PYTHON = "python"
    GO = "go"


class SessionStatus(StrEnum):
    NOT_STARTED = "not_started"
    IN_PROGRESS = "in_progress"
    SUBMITTED = "submitted"
    EXPIRED_SUBMITTED = "expired_submitted"


class ChatRole(StrEnum):
    CANDIDATE = "candidate"
    ASSISTANT = "assistant"
    SYSTEM = "system"


class ChatSource(StrEnum):
    TYPED = "typed"
    QUICK_PROMPT = "quick_prompt"
    QUESTION_HINT = "question_hint"
    CODE_REVIEW = "code_review"


class ChangeSource(StrEnum):
    """程式碼變更的來源。兩者 MUST NOT 混為一談（憲章原則 I）。"""

    CANDIDATE = "candidate"
    AI = "ai"


class EnvironmentEventType(StrEnum):
    WINDOW_BLUR = "window_blur"
    TAB_HIDDEN = "tab_hidden"


class Difficulty(StrEnum):
    EASY = "easy"
    MEDIUM = "medium"
    HARD = "hard"


class CamelModel(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)


# --- 對外實體 ---------------------------------------------------------------


class Example(CamelModel):
    input: str
    output: str
    note: str | None = None


class PublicQuestion(CamelModel):
    """刻意不含 predefined_tests：其內容 MUST NOT 出現在任何回應中（FR-030）。"""

    id: str
    title: str
    difficulty: Difficulty
    points: int
    description: str
    examples: list[Example]
    complexity_requirement: str
    grading_focus: list[str]
    starter_code: dict[str, str]
    quick_prompts: list[str]
    order: int
    test_count: int


class PublicSession(CamelModel):
    """僅含姓名與職稱兩項個資（FR-032）。新增欄位前 MUST 重新檢視個資最小化。"""

    id: str
    candidate_name: str
    position_title: str
    deadline_at: str | None
    status: SessionStatus


class PublicAnswer(CamelModel):
    question_id: str
    language: Language
    content: str
    saved_at: str
    revision: int


class PublicCodeBlock(CamelModel):
    block_index: int
    language: str | None
    content: str


class PublicChatMessage(CamelModel):
    id: str
    question_id: str
    role: ChatRole
    content: str
    created_at: str
    attached_code: str | None = None
    code_blocks: list[PublicCodeBlock] = Field(default_factory=list)


# --- 請求 body ---------------------------------------------------------------


class RedeemRequest(CamelModel):
    token: str = Field(min_length=1)


class SaveAnswerRequest(CamelModel):
    language: Language
    content: str
    revision: int = Field(ge=0)

    @field_validator("content")
    @classmethod
    def _within_limit(cls, value: str) -> str:
        if len(value.encode("utf-8")) > MAX_CONTENT_BYTES:
            raise ValueError("CONTENT_TOO_LARGE")
        return value


class SaveAnswerBatchItem(SaveAnswerRequest):
    question_id: str = Field(min_length=1)


class ApplyBlockRequest(CamelModel):
    """套用 AI 產出的程式碼區塊（FR-033）。"""

    message_id: str = Field(min_length=1)
    block_index: int = Field(ge=0)


class ChatRequest(CamelModel):
    question_id: str = Field(min_length=1)
    content: str = Field(min_length=1, max_length=8000)
    attach_code: bool = False
    source: ChatSource = ChatSource.TYPED


class ChatSystemRequest(CamelModel):
    from_question_id: str = Field(min_length=1)
    to_question_id: str = Field(min_length=1)


class EnvironmentEventItem(CamelModel):
    """MUST NOT 含任何判定性欄位（FR-026）——schema 沒有，就寫不進去。"""

    type: EnvironmentEventType
    started_at: str = Field(min_length=1)
    duration_ms: int = Field(ge=0)

File: backend/src/lib/schemas.ts
import { z } from 'zod';

/** 作答內容上限 256 KB（data-model.md：Answer 驗證規則）。 */
export const MAX_CONTENT_BYTES = 256 * 1024;

/** 環境事件門檻：短於 1000ms 的離開 MUST NOT 記錄（濾除焦點抖動）。 */
export const ENVIRONMENT_EVENT_MIN_DURATION_MS = 1000;

/** 剩餘時間低於此值時計時器轉為警示呈現（FR-020）。 */
export const TIMER_WARNING_THRESHOLD_SEC = 5 * 60;

export const languageSchema = z.enum(['javascript', 'typescript', 'python', 'go']);
export type Language = z.infer<typeof languageSchema>;

export const SUPPORTED_LANGUAGES: readonly Language[] = languageSchema.options;

export const sessionStatusSchema = z.enum([
  'not_started',
  'in_progress',
  'submitted',
  'expired_submitted',
]);
export type SessionStatus = z.infer<typeof sessionStatusSchema>;

export const guidanceModeSchema = z.enum(['light', 'deep']);
export type GuidanceMode = z.infer<typeof guidanceModeSchema>;

export const chatRoleSchema = z.enum(['candidate', 'assistant', 'system']);
export type ChatRole = z.infer<typeof chatRoleSchema>;

export const chatSourceSchema = z.enum(['typed', 'quick_prompt', 'question_hint', 'code_review']);
export type ChatSource = z.infer<typeof chatSourceSchema>;

export const environmentEventTypeSchema = z.enum(['window_blur', 'tab_hidden']);
export type EnvironmentEventType = z.infer<typeof environmentEventTypeSchema>;

export const difficultySchema = z.enum(['easy', 'medium', 'hard']);
export type Difficulty = z.infer<typeof difficultySchema>;

export const inviteTokenStatusSchema = z.enum(['pending', 'active', 'consumed', 'expired']);
export type InviteTokenStatus = z.infer<typeof inviteTokenStatusSchema>;

// --- 實體 -------------------------------------------------------------------

export const exampleSchema = z.object({
  input: z.string(),
  output: z.string(),
  note: z.string().optional(),
});
export type Example = z.infer<typeof exampleSchema>;

export const testCaseSchema = z.object({
  name: z.string(),
  expectedPass: z.boolean(),
});
export type TestCase = z.infer<typeof testCaseSchema>;

/**
 * 對外呈現的題目 —— 刻意不含 `predefinedTests`：
 * 其內容 MUST NOT 出現在任何回應中，避免應試者反推期望輸出（FR-030）。
 */
export const publicQuestionSchema = z.object({
  id: z.string(),
  title: z.string(),
  difficulty: difficultySchema,
  points: z.number().int(),
  description: z.string(),
  examples: z.array(exampleSchema),
  complexityRequirement: z.string(),
  gradingFocus: z.array(z.string()),
  starterCode: z.record(languageSchema, z.string()),
  quickPrompts: z.array(z.string()),
  order: z.number().int(),
  testCount: z.number().int().nonnegative(),
});
export type PublicQuestion = z.infer<typeof publicQuestionSchema>;

/**
 * 對外呈現的場次 —— 僅含姓名與職稱兩項個資（FR-032）。
 * 新增欄位前 MUST 確認不引入其他個人資料。
 */
export const publicSessionSchema = z.object({
  id: z.string(),
  candidateName: z.string(),
  positionTitle: z.string(),
  deadlineAt: z.string().nullable(),
  status: sessionStatusSchema,
  guidanceMode: guidanceModeSchema,
});
export type PublicSession = z.infer<typeof publicSessionSchema>;

export const publicAnswerSchema = z.object({
  questionId: z.string(),
  language: languageSchema,
  content: z.string(),
  savedAt: z.string(),
  revision: z.number().int(),
});
export type PublicAnswer = z.infer<typeof publicAnswerSchema>;

export const publicChatMessageSchema = z.object({
  id: z.string(),
  questionId: z.string(),
  role: chatRoleSchema,
  content: z.string(),
  createdAt: z.string(),
  attachedCode: z.string().nullable(),
});
export type PublicChatMessage = z.infer<typeof publicChatMessageSchema>;

// --- 請求 body ---------------------------------------------------------------

export const redeemRequestSchema = z.object({
  token: z.string().min(1),
});

const contentWithinLimit = z
  .string()
  .refine((v) => Buffer.byteLength(v, 'utf8') <= MAX_CONTENT_BYTES, {
    message: 'CONTENT_TOO_LARGE',
  });

export const saveAnswerRequestSchema = z.object({
  language: languageSchema,
  content: contentWithinLimit,
  revision: z.number().int().nonnegative(),
});
export type SaveAnswerRequest = z.infer<typeof saveAnswerRequestSchema>;

/** 離線補送：一次帶多筆，伺服端依 revision 排序套用。 */
export const saveAnswersBatchRequestSchema = z
  .array(saveAnswerRequestSchema.extend({ questionId: z.string().min(1) }))
  .min(1);
export type SaveAnswersBatchRequest = z.infer<typeof saveAnswersBatchRequestSchema>;

export const chatRequestSchema = z.object({
  questionId: z.string().min(1),
  content: z.string().min(1).max(8000),
  attachCode: z.boolean().default(false),
  source: chatSourceSchema.default('typed'),
});
export type ChatRequest = z.infer<typeof chatRequestSchema>;

export const chatSystemRequestSchema = z.object({
  fromQuestionId: z.string().min(1),
  toQuestionId: z.string().min(1),
});

export const guidanceModeRequestSchema = z.object({
  mode: guidanceModeSchema,
});

export const environmentEventsRequestSchema = z
  .array(
    z.object({
      type: environmentEventTypeSchema,
      startedAt: z.string().min(1),
      durationMs: z.number().int().nonnegative(),
    })
  )
  .min(1);
export type EnvironmentEventsRequest = z.infer<typeof environmentEventsRequestSchema>;
